File: frontend/lib/types/ml-operations.ts
/**
 * TypeScript types and Zod schemas for ML operations
 */

import { z } from "zod";

const numberWithDefault = (
  defaultValue: number
): z.ZodEffects<z.ZodNumber, number, unknown> =>
  z.preprocess((value) => value ?? defaultValue, z.number());

// =============================================================================
// Train Model Types
// =============================================================================

export const TrainModelRequestSchema = z.object({
  model_type: z
    .enum(["xgboost", "lightgbm", "random_forest", "dixon_coles", "hybrid_poisson"])
    .default("xgboost"),
  tune_hyperparameters: z.boolean().default(false),
  description: z.string().trim().min(1).max(500).optional(),
  activate: z.boolean().default(true),
  // Data range settings
  start_date: z.string().optional(),
  end_date: z.string().optional(),
  min_matches: z.number().min(10).max(100000).default(100),
  // Filtering: tournament and team selection
  tournament_ids: z.array(z.number()).optional(),
  team_ids: z.array(z.number()).optional(),
  venue_filter: z.enum(["home", "away", "both"]).optional(),
  // Match quality filters
  min_total_goals: z.number().min(0).optional(),
  max_total_goals: z.number().min(0).optional(),
  // Train/val/test split ratios
  train_ratio: z.number().min(0.1).max(0.9).default(0.7),
  val_ratio: z.number().min(0.05).max(0.45).default(0.15),
  test_ratio: z.number().min(0.05).max(0.45).default(0.15),
  // Training settings
  random_seed: z.number().min(0).max(999999).default(42),
  early_stopping_rounds: z.number().min(10).max(500).default(50),
  tuning_trials: z.number().min(10).max(500).default(50),
  // Calibration settings
  calibrate_probabilities: z.boolean().default(true),
  calibration_method: z
    .enum(["isotonic", "sigmoid", "temperature", "venn_abers"])
    .default("temperature"),
  // Outcome balancing
  outcome_balance: z.boolean().optional(),
  outcome_balance_strength: z.number().min(0).max(1).default(0.5),
  // Feature groups selection
  feature_groups: z.array(z.string()).optional(),
  // Feature importance pruning
  feature_selection: z.boolean().default(false),
  feature_selection_threshold: z.number().min(0).max(1).default(0.01),
  min_samples_per_feature: z.number().min(1).optional(),
  // Ensemble training
  use_ensemble: z.boolean().default(false),
  ensemble_types: z.array(z.string()).optional(),
  // Split strategy
  split_strategy: z
    .enum(["temporal", "expanding_window", "season_aware", "walk_forward"])
    .default("temporal"),
  gap_days: z.number().min(0).max(30).default(0),
  // Expanding window params
  min_train_size: z.number().min(50).max(5000).default(100),
  ew_val_size: z.number().min(10).max(1000).default(50),
  ew_test_size: z.number().min(10).max(1000).default(50),
  step_size: z.number().min(10).max(500).default(50),
  // Season-aware params
  train_seasons: z.number().min(1).max(10).default(3),
  val_seasons: z.number().min(1).max(5).default(1),
  test_seasons: z.number().min(1).max(5).default(1),
  // Model tags
  tags: z.record(z.string()).default({}),
  // Custom hyperparameters
  hyperparameters: z.record(z.unknown()).default({}),
});

export const TrainModelResultSchema = z.object({
  model_id: z.number().nullable(),
  model_version: z.string(),
  model_type: z.string(),
  is_active: z.boolean(),
  feature_schema_version: z.string(),
  num_features: z.number(),
  trained_at: z.string(),
  training_duration_seconds: z.number(),
  train_metrics: z.record(z.number()),
  val_metrics: z.record(z.number()),
  test_metrics: z.record(z.number()),
  feature_importance: z.record(z.number()).nullable(),
  // Ensemble metadata
  ensemble_weights: z.record(z.number()).nullable(),
  ensemble_validation_metrics: z.record(z.number()).nullable(),
  ensemble_types: z.array(z.string()).nullable(),
});

// =============================================================================
// Backtest Types
// =============================================================================

export const BacktestRequestSchema = z.object({
  model_version: z.string().optional(),
  tournament_id: z.number().optional(),
  start_date: z.string().optional(),
  end_date: z.string().optional(),
  min_matches: z.number().min(10).max(10000).default(100),
  min_edge: z.number().min(0).max(1).default(0).optional(),
});

export const ClassificationMetricsSchema = z.object({
  accuracy: z.number(),
  log_loss: z.number(),
  brier_score: z.number(),
  precision_macro: z.number(),
  recall_macro: z.number(),
  f1_macro: z.number(),
  precision_weighted: z.number(),
  recall_weighted: z.number(),
  f1_weighted: z.number(),
  per_class_precision: z.record(z.number()),
  per_class_recall: z.record(z.number()),
  per_class_f1: z.record(z.number()),
  confusion_matrix: z.array(z.array(z.number())),
  top_2_accuracy: z.number(),
  cohen_kappa: z.number(),
});

export const BettingMetricsSchema = z.object({
  total_bets: z.number(),
  winning_bets: z.number(),
  losing_bets: z.number(),
  total_stake: z.number(),
  total_return: z.number(),
  profit_loss: z.number(),
  roi_percent: z.number(),
  yield_percent: z.number(),
  sharpe_ratio: z.number(),
  max_drawdown: z.number(),
  win_rate: z.number(),
  average_winning_odds: z.number(),
  average_losing_odds: z.number(),
  average_kelly_fraction: z.number(),
  optimal_kelly_fraction: z.number(),
  mean_clv: numberWithDefault(0),
  clv_hit_rate: numberWithDefault(0),
  clv_weighted_roi: numberWithDefault(0),
});

export const BacktestResultSchema = z.object({
  model_version: z.string(),
  evaluated_at: z.string(),
  num_samples: z.number(),
  date_range: z.tuple([z.string(), z.string()]).nullable(),
  classification: ClassificationMetricsSchema,
  betting: BettingMetricsSchema.nullable(),
  expected_calibration_error: z.number(),
  maximum_calibration_error: z.number(),
  outcome_accuracy: z.record(z.number()),
});

// =============================================================================
// Calibrate Types
// =============================================================================

export const CalibrateRequestSchema = z.object({
  model_version: z.string().optional(),
  method: z
    .enum(["isotonic", "sigmoid", "temperature", "venn_abers"])
    .default("isotonic"),
  validation_split: z.number().min(0.1).max(0.5).default(0.2),
  activate: z.boolean().default(true),
});

export const CalibrationMetricsSchema = z.object({
  expected_calibration_error: z.number(),
  maximum_calibration_error: z.number(),
  brier_score: z.number(),
  log_loss: z.number(),
});

export const CalibrateResultSchema = z.object({
  base_model_version: z.string(),
  calibrated_model_version: z.string(),
  method: z.string(),
  samples_used: z.number(),
  before_metrics: CalibrationMetricsSchema,
  after_metrics: CalibrationMetricsSchema,
  improvement: z.object({
    ece_improvement: z.number(),
    brier_improvement: z.number(),
    log_loss_improvement: z.number(),
  }),
  is_active: z.boolean(),
});

// =============================================================================
// TypeScript Interfaces
// =============================================================================

export type TrainModelRequest = z.infer<typeof TrainModelRequestSchema>;
export type TrainModelResult = z.infer<typeof TrainModelResultSchema>;

export type BacktestRequest = z.infer<typeof BacktestRequestSchema>;
export type ClassificationMetrics = z.infer<typeof ClassificationMetricsSchema>;
export type BettingMetrics = z.infer<typeof BettingMetricsSchema>;
export type BacktestResult = z.infer<typeof BacktestResultSchema>;

export type CalibrateRequest = z.infer<typeof CalibrateRequestSchema>;
export type CalibrationMetrics = z.infer<typeof CalibrationMetricsSchema>;
export type CalibrateResult = z.infer<typeof CalibrateResultSchema>;

// =============================================================================
// Backtest History Types
// =============================================================================

export const BacktestHistoryItemSchema = z.object({
  id: z.number(),
  model_version_id: z.number(),
  model_name: z.string().nullable(),
  model_version: z.string().nullable(),
  num_samples: z.number(),
  date_range_start: z.string().nullable(),
  date_range_end: z.string().nullable(),
  accuracy: z.number(),
  log_loss: z.number(),
  f1_macro: z.number(),
  roi_percent: z.number().nullable(),
  win_rate: z.number().nullable(),
  evaluated_at: z.string(),
});

export const BacktestHistoryListSchema = z.object({
  items: z.array(BacktestHistoryItemSchema),
  total: z.number(),
});

export type BacktestHistoryItem = z.infer<typeof BacktestHistoryItemSchema>;
export type BacktestHistoryList = z.infer<typeof BacktestHistoryListSchema>;

File: algobet/api/schemas/ml_operations.py
"""Request and response schemas for ML operation endpoints."""

from __future__ import annotations

import os
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

DEFAULT_TRAIN_MODEL_TYPE = os.getenv("ALGOBET_DEFAULT_MODEL_TYPE", "xgboost")


class CalibrateResultResponse(BaseModel):
    """Response schema for calibrate results."""

    base_model_version: str
    calibrated_model_version: str
    method: str
    samples_used: int
    before_metrics: CalibrationMetricsResponse
    after_metrics: CalibrationMetricsResponse
    improvement: dict[str, float]
    is_active: bool


class BacktestHistoryItem(BaseModel):
    """Backtest history item for list response."""

    id: int
    model_version_id: int
    model_name: str | None = None
    model_version: str | None = None
    num_samples: int
    date_range_start: str | None
    date_range_end: str | None
    accuracy: float
    log_loss: float
    f1_macro: float
    roi_percent: float | None
    win_rate: float | None
    evaluated_at: str


class CalibrationMetricsResponse(BaseModel):
    """Calibration metrics response."""

    expected_calibration_error: float
    maximum_calibration_error: float
    brier_score: float
    log_loss: float


class BacktestRequest(BaseModel):
    """Request schema for backtest operation."""

    model_version: str | None = None
    tournament_id: int | None = None
    start_date: datetime | None = None
    end_date: datetime | None = None
    min_matches: int = Field(default=100, ge=10, le=10000)
    min_edge: float = Field(default=0.0, ge=0.0, le=1.0)


class BacktestHistoryListResponse(BaseModel):
    """Response for backtest history list."""

    items: list[BacktestHistoryItem]
    total: int


class BettingMetricsResponse(BaseModel):
    """Betting metrics response."""

    total_bets: int
    winning_bets: int
    losing_bets: int
    total_stake: float
    total_return: float
    profit_loss: float
    roi_percent: float
    yield_percent: float
    sharpe_ratio: float
    max_drawdown: float
    win_rate: float
    average_winning_odds: float
    average_losing_odds: float
    average_kelly_fraction: float
    optimal_kelly_fraction: float
    mean_clv: float = 0.0
    clv_hit_rate: float = 0.0
    clv_weighted_roi: float = 0.0


class BacktestResultResponse(BaseModel):
    """Response schema for backtest results."""

    model_version: str
    evaluated_at: str
    num_samples: int
    date_range: tuple[str, str] | None
    classification: ClassificationMetricsResponse
    betting: BettingMetricsResponse | None = None
    expected_calibration_error: float
    maximum_calibration_error: float
    outcome_accuracy: dict[str, float]


class CalibrateRequest(BaseModel):
    """Request schema for calibrate operation."""

    model_version: str | None = None
    method: str = Field(
        default="isotonic", pattern="^(isotonic|sigmoid|temperature|venn_abers)$"
    )
    validation_split: float = Field(default=0.2, ge=0.1, le=0.5)
    activate: bool = True


class ClassificationMetricsResponse(BaseModel):
    """Classification metrics response."""

    accuracy: float
    log_loss: float
    brier_score: float
    precision_macro: float
    recall_macro: float
    f1_macro: float
    precision_weighted: float
    recall_weighted: float
    f1_weighted: float
    per_class_precision: dict[str, float]
    per_class_recall: dict[str, float]
    per_class_f1: dict[str, float]
    confusion_matrix: list[list[int]]
    top_2_accuracy: float
    cohen_kappa: float


class TrainModelRequest(BaseModel):
    """Request schema for model training."""

    model_type: str = Field(
        default=DEFAULT_TRAIN_MODEL_TYPE,
        pattern="^(xgboost|lightgbm|random_forest|dixon_coles|hybrid_poisson|catboost)$",
    )
    tune_hyperparameters: bool = False
    description: str | None = None
    activate: bool = True

    # Data range settings
    start_date: datetime | None = None
    end_date: datetime | None = None
    min_matches: int = Field(default=100, ge=10, le=100000)

    # Filtering: tournament and team selection
    tournament_ids: list[int] | None = None
    team_ids: list[int] | None = None
    venue_filter: str | None = None  # "home", "away", "both"

    # Match quality filters
    min_total_goals: float | None = None
    max_total_goals: float | None = None

    # Train/val/test split ratios
    train_ratio: float = Field(default=0.7, ge=0.1, le=0.9)
    val_ratio: float = Field(default=0.15, ge=0.05, le=0.45)
    test_ratio: float = Field(default=0.15, ge=0.05, le=0.45)

    # Training settings
    random_seed: int = Field(default=42, ge=0, le=999999)
    early_stopping_rounds: int = Field(default=50, ge=10, le=500)
    tuning_trials: int = Field(default=50, ge=10, le=500)

    # Calibration settings
    calibrate_probabilities: bool = True
    calibration_method: str = Field(
        default="temperature", pattern="^(isotonic|sigmoid|temperature|venn_abers)$"
    )

    # Outcome balancing control
    outcome_balance: bool | None = None
    outcome_balance_strength: float = Field(default=0.5, ge=0.0, le=1.0)

    # Feature importance pruning
    feature_selection: bool = True
    feature_selection_threshold: float = Field(default=0.01, ge=0.0, le=1.0)
    min_samples_per_feature: int | None = Field(default=None, ge=1)

    # Feature groups selection (subset of available generators)
    feature_groups: list[str] | None = None

    # Ensemble training
    use_ensemble: bool = False
    ensemble_types: list[str] | None = None  # e.g. ["xgboost", "lightgbm"]

    # Stacking ensemble
    use_stacking_ensemble: bool = False
    stacking_base_models: list[str] = Field(
        default_factory=lambda: ["xgboost", "dixon_coles"]
    )
    stacking_meta_learner: str = Field(default="logistic", pattern="^(logistic|mlp)$")
    stacking_n_folds: int = Field(default=5, ge=3, le=10)

    # Split strategy
    split_strategy: str = Field(
        default="walk_forward",
        pattern="^(temporal|expanding_window|season_aware|walk_forward)$",
    )
    gap_days: int = Field(default=0, ge=0, le=30)
    # Expanding window params (used when split_strategy = expanding_window)
    min_train_size: int = Field(default=100, ge=50, le=5000)
    ew_val_size: int = Field(default=50, ge=10, le=1000)
    ew_test_size: int = Field(default=50, ge=10, le=1000)
    step_size: int = Field(default=50, ge=10, le=500)
    # Season-aware params (used when split_strategy = season_aware)
    train_seasons: int = Field(default=8, ge=1, le=10)
    val_seasons: int = Field(default=1, ge=1, le=5)
    test_seasons: int = Field(default=1, ge=1, le=5)

    # Model tags
    tags: dict[str, str] = Field(default_factory=dict)

    # Custom hyperparameters (optional)
    hyperparameters: dict[str, Any] = Field(default_factory=dict)


class TrainModelResponse(BaseModel):
    """Response schema for model training."""

    model_id: int | None = None
    model_version: str
    model_type: str
    is_active: bool
    feature_schema_version: str
    num_features: int
    trained_at: str
    training_duration_seconds: float
    train_metrics: dict[str, float]
    val_metrics: dict[str, float]
    test_metrics: dict[str, float]
    feature_importance: dict[str, float] | None = None
    # Ensemble metadata
    ensemble_weights: dict[str, float] | None = None
    ensemble_validation_metrics: dict[str, float] | None = None
    ensemble_types: list[str] | None = None
    # Stacking metadata
    stacking_metadata: dict[str, Any] | None = None
    # Calibration metadata
    calibration_metadata: dict[str, Any] | None = None
